### PR TITLE
ZeroCopy likely subtags

### PR DESCRIPTION
**Pull request by robertbastian**
#1034, #1488

### CODE REVIEW

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs [hunk 1]
```diff
@@ -72,62 +72,77 @@ impl IterableResourceProvider<LikelySubtagsV1Marker> for LikelySubtagsProvider {
...
-                    script_region.insert((scr.into(), reg.into()), extract_result(entry));
+                if let Some(region) = entry.0.region {
+                    with_diff!((language, None, None) => script_region.insert(&(scr, region), &language));
                 } else {
-                    script.insert(scr.into(), extract_result(entry));
+                    with_diff!((language, None, Some(region)) => script.insert(&scr, &(language, region)));
                 }
             } else if let Some(reg) = entry.0.region {
-                region.insert(reg.into(), extract_result(entry));
+                // Some of the target regions here are not equal to the source, such as und-002 -> en-Latn-NG.
```
Comment (Manishearth) [inline provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:137]: Calling this bit out for @zbraniecki or @sffc

**Review by Manishearth — DISMISSED** [review]
Looks good from my side, but would like someone to look at the code that's discarding regions

**Review by Manishearth — APPROVED** [review]

File: components/locale_canonicalizer/src/provider.rs [hunk 2]
```diff
@@ -82,30 +85,29 @@ pub struct AliasesV1<'data> {
...
+/// for searching and replacing. E.g., the `language_script` field is used to store
+/// rules for `LanguageIdentifier`s that contain a language and a script, but not a
+/// region.
+#[yoke(prove_covariance_manually)]
+pub struct LikelySubtagsV1<'data> {
     /// Language and script.
-    #[zerofrom(clone)]
-    pub language_script: LiteMap<(TinyAsciiStr<3>, TinyAsciiStr<4>), LanguageIdentifier>,
+    #[cfg_attr(feature = "serialize", serde(borrow))]
+    pub language_script: ZeroMap<'data, (Language, Script), Region>,
```
Comment (sffc) [inline components/locale_canonicalizer/src/provider.rs:95]: Discussion: For lookup (the left side of the map), we can use "raw" types, because we don't need to validate them. We only need to validate things if they are used for retrieval (the right side of the map).

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs (hunk 1 above)
Comment (sffc) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:137]: For for those entries, like "und-002 -> en-Latn-NG", does the current code have the same behavior as the new code?

File: components/locale_canonicalizer/src/provider.rs (hunk 2 above)
Comment (robertbastian) [reply components/locale_canonicalizer/src/provider.rs]: Good point. We should also update the value types in `AliasesV1` to be validated.

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:137]: Yes, a subtag only gets updated if it's non-empty: https://github.com/unicode-org/icu4x/blob/main/components/locale_canonicalizer/src/locale_canonicalizer.rs#L202

I'm a bit sceptical about the correctness of that though, if CLDR gives us this data, maybe the reference algorithm doesn't actually discard it.

**Review by sffc — APPROVED** [review]

**Review by Manishearth — DISMISSED** [review]

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs [hunk 3]
```diff
@@ -72,62 +72,77 @@ impl IterableResourceProvider<LikelySubtagsV1Marker> for LikelySubtagsProvider {
...
         for entry in other.supplemental.likely_subtags.iter() {
+            // Computes the delta of the entry and assigns to the pattern.
+            // Errors if the delta is not assignable to the pattern.
+            macro_rules! with_diff {
+                ($pat:pat => $stmt:expr ) => {
+                    if let $pat = (
+                        if entry.0.language != entry.1.language {
+                            entry.1.language
+                        } else {
+                            Language::UND
```
Comment (zbraniecki) [inline provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:96]: `question`: why `UND` instead of `Option<Language>` and None here?

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs (hunk 1 above)
Comment (zbraniecki) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:137]: I'm ok with this for now. We may want to dwell into this detail if we hit a case where it matters and it shouldn't be hard to fine-tune the code.

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs (hunk 3 above)
Comment (robertbastian) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:96]: Because `UND` is our (and BCP-47's) representation of an absent language code. If we used `Option<Language>`, absence would have two values (`Some(Language::UND)` and `None`).

File: components/locale_canonicalizer/src/provider.rs [hunk 4]
```diff
@@ -7,12 +7,21 @@
...
 use icu_provider::prelude::*;
 use litemap::LiteMap;
 use tinystr::TinyAsciiStr;
 use zerovec::{ZeroMap, ZeroSlice};
 
+type UnvalidatedLanguage = TinyAsciiStr<3>;
+type UnvalidatedScript = TinyAsciiStr<4>;
+type UnvalidatedRegion = TinyAsciiStr<3>;
+type UnvalidatedVariant = TinyAsciiStr<8>;
+type UnvalidatedValue = TinyAsciiStr<7>;
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/locale_canonicalizer/src/provider.rs:23]: Done

File: provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs (hunk 3 above)
Comment (robertbastian) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:96]: This is consistent with the three (stack) fields of `LanguageIdentifier`

Comment (zbraniecki) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:96]: The reason we do this on `LanguageIdentifier` is that LI *cannot* exist without language. The "empty" language is therefore just one of the languages rather than representation of lack of language.
It is intended to help API users understand the difference between omitting script/region and omitting language.

In your code that is not important. You navigate on the LSR triplet and communicate which of the three are present. I think `(Option<Language>, Option<Script>, Option<Region>)` is the correct type for it.

Comment (zbraniecki) [reply provider/datagen/src/cldr/transform/locale_canonicalizer/likely_subtags.rs:96]: Ah, that's a good point! 
Thank you for taking time to explain it to me. I'm convinced.

**Review by zbraniecki — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]
there are some test failures

**Comment by robertbastian** [comment]
Yes I'm working on debugging that, `icu_testdata::get_provider` works, `icu_testdata::get_static_provider` returns a very helpful `"Serde error"`

**Comment by Manishearth** [comment]
very helpful indeed

**Comment by sffc** [comment]
> Yes I'm working on debugging that, `icu_testdata::get_provider` works, `icu_testdata::get_static_provider` returns a very helpful `"Serde error"`

Try to enable logging to get more helpful output. Background: #1449

**Review by Manishearth — DISMISSED** [review]

File: components/locid/Cargo.toml [hunk 5]
```diff
@@ -46,6 +46,7 @@ icu = { path = "../icu", default-features = false }
 icu_benchmark_macros = { version = "0.5", path = "../../tools/benchmark/macros" }
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
+postcard = { version = "0.7.0", features = ["use-std"] }
```
Comment (sffc) [inline components/locid/Cargo.toml:49]: Nit: Apply `default-features = false` whenever we link `postcard` because we want to avoid the `heapless-cas` feature being pulled in because it contains undesirable dependencies. It's less of a problem for `dev-dependencies` but still good practice.

File: components/locid/src/serde.rs [hunk 6]
```diff
@@ -65,7 +65,12 @@ impl<'de> Deserialize<'de> for Language {
             }
         }
 
-        deserializer.deserialize_string(LanguageVisitor)
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_string(LanguageVisitor)
+        } else {
+            Self::try_from_raw(Deserialize::deserialize(deserializer)?)
+                .map_err(serde::de::Error::custom)
+        }
```
Comment (sffc) [inline components/locid/src/serde.rs:73]: Praise: This is a good change.

Nit: Could be in a separate PR since it seems to be independent from the rest of the PR.

**Review by sffc — APPROVED** [review]

File: components/locid/Cargo.toml (hunk 5 above)
Comment (robertbastian) [reply components/locid/Cargo.toml]: Done

File: components/locid/src/serde.rs (hunk 6 above)
Comment (robertbastian) [reply components/locid/src/serde.rs:73]: It's very dependent, I'm adding this code

**Review by Manishearth — APPROVED** [review]

**Review by sffc — APPROVED** [review]
